src/economy/economy.cpp

```diff
@@ -143,80 +712,145 @@ bool has_factory(sys::state const& state, dcon::state_instance_id si) {
 	{
 		state.world.for_each_commodity([&](dcon::commodity_id c) {
-			auto kf = state.world.commodity_get_key_factory(c);
-			if(state.world.commodity_get_is_life_need(c) && (state.world.commodity_get_is_available_from_start(c) || (kf && state.world.nation_get_active_building(n, kf)))) {
-				auto& w = state.world.nation_get_life_needs_weights(n, c);
-				w = 1.f;
+			if(valid_life_need(state, nation, c)) {
+				auto& w = state.world.market_get_life_needs_weights(n, c);
+				w = 0.001f;
 			}
 		});
 	}
 	{
 		state.world.for_each_commodity([&](dcon::commodity_id c) {
 			auto kf = state.world.commodity_get_key_factory(c);
-			if(state.world.commodity_get_is_everyday_need(c) && (state.world.commodity_get_is_available_from_start(c) || (kf && state.world.nation_get_active_building(n, kf)))) {
-				auto& w = state.world.nation_get_everyday_needs_weights(n, c);
-				w = 1.f;
+			if(valid_everyday_need(state, nation, c)) {
+				auto& w = state.world.market_get_everyday_needs_weights(n, c);
+				w = 0.001f;
 			}
 		});
 	}
 	{
 		state.world.for_each_commodity([&](dcon::commodity_id c) {
 			auto kf = state.world.commodity_get_key_factory(c);
-			if(state.world.commodity_get_is_luxury_need(c) && (state.world.commodity_get_is_available_from_start(c) || (kf && state.world.nation_get_active_building(n, kf)))) {
-				auto& w = state.world.nation_get_luxury_needs_weights(n, c);
-				w = 1.f;
+			if(valid_luxury_need(state, nation, c)) {
+				auto& w = state.world.market_get_luxury_needs_weights(n, c);
+				w = 0.001f;
 			}
 		});
 	}
 }
 
-float need_weight(sys::state& state, dcon::nation_id n, dcon::commodity_id c) {
-	return 1.0f / math::sqrt(std::max(state.world.commodity_get_current_price(c), 0.001f));
+float need_weight(sys::state& state, dcon::market_id n, dcon::commodity_id c) {
+	return 1.0f / std::max(price(state, n, c), 0.0001f);
 }
 
-void rebalance_needs_weights(sys::state& state, dcon::nation_id n) {
+void rebalance_needs_weights(sys::state& state, dcon::market_id n) {
+	auto zone = state.world.market_get_zone_from_local_market(n);
+	auto nation = state.world.state_instance_get_nation_from_state_ownership(zone);
+
 	{
-		float total_weights = 0.0f;
+		float total_weights = 0.f;
 		uint32_t count = 0;
 
 		state.world.for_each_commodity([&](dcon::commodity_id c) {
-			auto kf = state.world.commodity_get_key_factory(c);
-			if(state.world.commodity_get_is_life_need(c) && (state.world.commodity_get_is_available_from_start(c) || (kf && state.world.nation_get_active_building(n, kf)))) {
+			if(valid_life_need(state, nation, c)) {
 				auto weight = need_weight(state, n, c);
 				total_weights += weight;
 				count++;
 			}
 		});
 
+		// remove some of needs completely depending on price
+		float total_weights_adjusted = 0.f;
+		state.world.for_each_commodity([&](dcon::commodity_id c) {
+			if(valid_life_need(state, nation, c)) {
+				auto weight = need_weight(state, n, c);
+				if(weight > total_weights * 0.05f) {
```

Comment:
What happens if there are enough life need items that *all* of them being less than 5% is a possible occurrence?